### ISSUE TITLE
Question about message size within actix-raft (and Raft in general).

### ISSUE THREAD

**Issue by joshgev** [railgun-rs/actix-raft#56]
Hi everyone!  If this is not an appropriate venue for the following question, please let me know.

TLDR: Is there a limit (hard or otherwise) to the size that Raft messages for committing log updates can be?  I want to write a Raft application using `actix-raft` whose log updates are anywhere from a few bytes to hundreds of **megabytes** large; is this possible?

Long version:

I'd like to implement a simple file duplication system to keep replicate a folder across several different computers.  For context: standard network/distributed file systems like `nfs` and `gluster` almost work for my use case but fail in a critical way: they are too slow.  My application requires shared files to be read quickly by software running on the nodes that comprise my distributed system.  (Actually `gluster` fails even more because of some of the shortcomings associated with user-space filesystems).  

I've been looking to implement a pretty simple file replication system that just makes sure a given folder is appropriately synced up.  Consensus is pretty important because if a file is visible to node A, then it must be visible to the other nodes otherwise we get into trouble (replication doesn't have to be immediate, but it should happen eventually).  As a consequence, I've been looking at Raft for a few days and I am pretty eager to try a toy implementation with `actix-raft`.  I believe I have a handle on most aspects of the project, but I have one lingering question that I can't seem to find an answer to online.  

The file sizes will range from a few bytes to a few hundred megabytes.  The way I see this system working is that if I write a file to any of the synced directories, the replicator running for that directory will sense a change (using `inotify`) and will use Raft to spread the new file around and reach a consensus so that, eventually, all functioning replicators (this is what I am terming the software that will use `actix-raft`) on the various nodes will have picked up the change and written the file locally.  Now, without the use of some third-party storage mechanism (e.g. Azure/Azurite, S3/Minio, etc), it seems that to make this work with Raft, the messages that Raft uses internally to form/disseminate consensus will have to contain the relevant files themselves.   My questions then are:

1. Have I misunderstood anything about the appropriate use of Raft in the scenario I've described
2. Are there hard limits on message size?  Do the semantics of Raft allow or disallow what I am proposing?

Thanks in advance for any guidance here.

**Comment by thedodd** [comment]
Hey @joshgev great question! Sounds like an awesome use case. 

As far as your questions, I do have a few concerns:
- raft is definitely more geared towards consensus over changes in state to some underlying state machine. In your use case, your state machine is more like an entire filesystem, or at least some portion of it.
- this is going to be tricky, because raft messages are intended to hold the entire contents of the message. Messages are pipelines and synced to all follower nodes. If you are dealing with 100Mb sized messages ... well you can imagine how much memory you would be dealing with very early on.

Based on the above observations, my recommendation would be that you have a separate mechanism for being able to capture the file changes taking place on a node, probably serializing those changes into an easily identifiable hash; and then Raft can simply pass around those hashes.

A pattern like that would maximize Raft's strengths, and would allow you to achieve a decoupled architecture. What you could do is simply have the nodes expose a mechanism outside of raft for actually streaming the contents of the files between nodes, but identify them (and their version) based on the hash, which would be passed around via the Raft messages.

To be more specific, the Raft leader will send replication requests to follower nodes instructing them to sync/delete/whatever a specific file by hash ID or whatever. The storage engine implementation of yours can just sync those files by hashID as part of the "update state machine" lifecycle. Once they've properly synced the file (or whatever action they are supposed to take), then they can respond to the Raft leader telling them that they are finished.

So, to state this succinctly, I think having large Raft RPC message sizes will be a bad idea long-term. You will hit memory scalability issues rather quickly, I would imagine. Instead, use hashes or something like that to identify the large data chunks.

Thoughts?

**Comment by joshgev** [comment]
Thanks a ton for your reply, @thedodd.  I fully expected your answer, though I had hoped you'd say "sure, that's totally fine."  Your explanation is clear and I can modify the image in my head to include an external storage mechanism.
